rust/repositories.bzl

```diff
@@ -42,98 +78,333 @@ filegroup(
+        tool_subdirectory = "rustc",
+        version = ctx.attr.version,
+    )
+
+    compiler_BUILD = BUILD_for_compiler(target_triple)
+
+    return compiler_BUILD
+
+def _load_rust_stdlib(ctx, target_triple):
+    """Loads a rust standard library and yields corresponding BUILD for it
+
+    Args:
+      ctx: A repository_ctx.
+      target_triple: The rust-style target triple of the tool
+    Returns:
+      The BUILD file contents for this stdlib, and a toolchain decl to match
+    """
+
+    load_arbitrary_tool(
+        ctx,
+        iso_date = ctx.attr.iso_date,
+        param_prefix = "rust-std_",
+        target_triple = target_triple,
+        tool_name = "rust-std",
+        tool_subdirectory = "rust-std-{}".format(target_triple),
+        version = ctx.attr.version,
     )
 
-def _default_toolchains():
-    all_toolchains = [
-        ("rust-linux-x86_64", "x86_64-unknown-linux-gnu"),
-        ("rust-darwin-x86_64", "x86_64-apple-darwin"),
-        ("rust-freebsd-x86_64", "x86_64-unknown-freebsd"),
+    toolchain_prefix = ctx.attr.toolchain_name_prefix or DEFAULT_TOOLCHAIN_NAME_PREFIX
+    stdlib_BUILD = BUILD_for_stdlib(target_triple)
+    toolchain_BUILD = BUILD_for_rust_toolchain(
+        name = "{toolchain_prefix}_{target_triple}".format(
+            toolchain_prefix = toolchain_prefix,
+            target_triple = target_triple,
+        ),
+        exec_triple = ctx.attr.exec_triple,
+        target_triple = target_triple,
+        workspace_name = ctx.attr.name,
+    )
+
+    return stdlib_BUILD + toolchain_BUILD
+
+def _rust_toolchain_repository_impl(ctx):
+    """The implementation of the rust toolchain repository rule."""
+
+    _check_version_valid(ctx.attr.version, ctx.attr.iso_date)
+
+    BUILD_components = [
+        _load_rust_compiler(ctx),
+        _load_rust_stdlib(ctx, ctx.attr.exec_triple),
     ]
 
-    all_toolchain_BUILDs = []
-    for toolchain in all_toolchains:
-        all_toolchain_BUILDs.append(BUILD_for_toolchain(toolchain[0], toolchain[1]))
+    for extra_stdlib_triple in ctx.attr.extra_target_triples:
+        BUILD_components.append(_load_rust_stdlib(ctx, extra_stdlib_triple))
 
-    return """
-load("@io_bazel_rules_rust//rust:toolchain.bzl", "rust_toolchain")
+    ctx.file("WORKSPACE", "")
+    ctx.file("BUILD", "\n".join(BUILD_components))
+
+def _rust_toolchain_repository_proxy_impl(ctx):
+    BUILD_components = [
+        BUILD_for_toolchain(
+            name = "{toolchain_prefix}_{target_triple}".format(
+                toolchain_prefix = ctx.attr.toolchain_name_prefix,
+                target_triple = ctx.attr.exec_triple,
+            ),
+            exec_triple = ctx.attr.exec_triple,
+            parent_workspace_name = ctx.attr.parent_workspace_name,
+            target_triple = ctx.attr.exec_triple,
+        ),
+    ]
 
-{all_toolchain_BUILDs}
-""".format(all_toolchain_BUILDs = "\n".join(all_toolchain_BUILDs))
+    for target_triple in ctx.attr.extra_target_triples:
```

Comment:
can dedupe

for target_triple in [ctx.attr.exec_triple] + ctx.attr.extra_target_triples